File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
from collections import defaultdict

def solution(want, number, discount):
    # 10일 연속으로 종류, 수량이 일치해야 함
    # 완탐 10만 * 조건 확인 10 = 100만
    
    # 10일동안 모두 구할 수 있다면 카운팅, 
    # 첫 10일은 구하고 그 다음은 이전 수량 목록을 이용
    # 새로운 것 추가하고 가장 오래된 것 버리기
    
    # 목표 수량
    target_dict = defaultdict(int)
    for w,n in zip(want,number):
        target_dict[w] = n
    # 결과 카운팅
    cnt = 0
    # dict로 비교
    dict = defaultdict(int)
    for t in discount[:10]:
        dict[t] += 1
    correct = True
    for key in dict: # 하나라도 일치하지 않으면 False
        if target_dict[key] != dict[key]:
            correct = False
            break
    if correct: # 모두 일치해야 카운팅
        cnt += 1
    for i in range(10,len(discount)):
        dict[discount[i]] += 1
        dict[discount[i-10]] -= 1
        correct = True
        for key in dict: # 하나라도 일치하지 않으면 False
            if target_dict[key] != dict[key]:
                correct = False
                break
        if correct: # 모두 일치해야 카운팅
            cnt += 1
    return cnt
        
        
        
        

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public static void log(Object o){
        System.out.println(o);
    }
    public int solution(String[] want, int[] number, String[] discount) {
        /*
        최초에 want*number의 길이 만큼 discount를 돌면서
        count를 감소시킨다. (want와 무관한 것이면 pass)
        만약 count 중 0보다 큰 것이 있다면 해당 다음날로 넘어간다.
        하루씩 이동하면서 start는 빼고 end를 더한다.
        가능한 날이 없으면 0을 반환한다.
        */
        Map<String, Integer> count = new HashMap<>();
        // count 초기화
        int n = want.length;
        int total = 0;
        for(int i=0; i<n; i++){
            String key = want[i];
            Integer value = number[i];
            total += value;
            count.put(key, value);
        }
        //log(count);
        
        int result = 0;
        // 첫 시뮬
        for(int i=0; i<total; i++){
            String key = discount[i];
            if(count.containsKey(key)){
                int cnt = count.get(key);
                count.put(key, cnt - 1);
            }
        }
        boolean zero = true;
        for(String key: count.keySet()){
            if(count.get(key) != 0){
                zero = false;
                break;
            }
        }
        if(zero){
            result += 1;
        }
        for(int i=0; i+total<discount.length; i++){
            String startKey = discount[i];
            String endKey =  discount[i+total];
            if(count.containsKey(startKey)){
                count.put(startKey, count.get(startKey) + 1);
            }
            if(count.containsKey(endKey)){
                count.put(endKey, count.get(endKey) - 1);
            }
            // 검증
            zero = true;
            for(String key: count.keySet()){
                if(count.get(key) != 0){
                    zero = false;
                    break;
                }
            } 
            //log(count);
            if(zero){
                result += 1;
            }
        }
        return result;
    }
}